Implement lower and upper bound inference in function pointer parameters

333fred opened the pull request:
https://github.com/dotnet/roslyn/pull/43041 implemented exact inference of function pointer parameters and bounds only, and left a few errors in our tests that needed to be handled. This is also the cause of https://github.com/dotnet/roslyn/issues/50096. This implements bounds inference, using the same variance rules that function pointer to function pointer conversions do. Note that for parameters and return types, I'm not considering delegate*->void* conversions, as `T` cannot be either `void` or `void*`.

Spec PR: https://github.com/dotnet/csharplang/pull/4310

Fixes https://github.com/dotnet/roslyn/issues/50096.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1563,7 +1667,6 @@ public static void M()
     }
 }");
 
-            // Some of these errors should become CS0306 after variant conversions are implemented, tracked by https://github.com/dotnet/roslyn/issues/39865
```
None of the remaining 0411 errors can be inferred: they're between function pointers with int and string parameters. There is no reference conversion or implicit pointer conversion, so nothing can be inferred.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1651,7 +1754,6 @@ public static void M()
     }
 }");
 
-            // These should all work: https://github.com/dotnet/roslyn/issues/39865
```
These could only be inferred if we supported array covariance for implicit pointer conversions. The spec makes no mention of such support, and we do not support `int*[]` being converted to `void*[]` today, so this is not expected to be supported.

333fred commented:
@dotnet/roslyn-compiler for review.

333fred commented:
@dotnet/roslyn-compiler for review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -501,10 +501,10 @@ private bool HasBound(int methodTypeParameterIndex)
         }
 
-        private NamedTypeSymbol GetFixedDelegate(NamedTypeSymbol delegateType)
+        private TypeSymbol GetFixedDelegateOrFunctionPointer(TypeSymbol delegateType)
```
>delegateType [](start = 72, length = 12)

No longer a delegate type. Consider renaming as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -861,26 +861,26 @@ private static bool DoesInputTypeContain(BoundExpression argument, TypeSymbol fo
             }
 
-            if (argument.Kind != BoundKind.UnboundLambda && argument.Kind != BoundKind.MethodGroup)
+            if (argument.Kind is not (BoundKind.UnboundLambda or BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator))
```
>if (argument.Kind is not (BoundKind.UnboundLambda or BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator)) [](start = 12, length = 118)

It doesn't feel right to continue with ```UnconvertedAddressOfOperator``` for a delegate and with ```UnboundLambda``` for a function pointer.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -915,30 +915,36 @@ private bool HasUnfixedParamInInputType(BoundExpression pSource, TypeSymbol pDes
             }
 
-            if (argument.Kind != BoundKind.UnboundLambda && argument.Kind != BoundKind.MethodGroup)
+            if (argument.Kind is not (BoundKind.UnboundLambda or BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator))
```
>if (argument.Kind is not (BoundKind.UnboundLambda or BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator)) [](start = 12, length = 118)

Same comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -1281,60 +1287,73 @@ private bool MethodGroupReturnTypeInference(Binder binder, BoundExpression sourc
             // SPEC:   inference is made from U to Tb.
 
-            if (source.Kind != BoundKind.MethodGroup)
+            if (source.Kind is not (BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator))
```
>if (source.Kind is not (BoundKind.MethodGroup or BoundKind.UnconvertedAddressOfOperator)) [](start = 12, length = 89)

Similar comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2067,6 +2091,61 @@ private void LowerBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+                var sourceParam = sourceSignature.Parameters[i];
+                var targetParam = targetSignature.Parameters[i];
+
+                if ((sourceParam.Type.IsReferenceType || sourceParam.Type.IsFunctionPointer()) && sourceParam.RefKind == RefKind.None)
```
>sourceParam.Type.IsFunctionPointer() [](start = 57, length = 36)

Does this match the specification?

333fred replied:
I believe I've updated the specification to match this :).

---
In reply to: [554276075](https://github.com/dotnet/roslyn/pull/50249#discussion_r554276075) [](ancestors = 554276075)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2067,6 +2091,61 @@ private void LowerBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+                }
+            }
+
+            if ((sourceSignature.ReturnType.IsReferenceType || sourceSignature.ReturnType.IsFunctionPointer()) && sourceSignature.RefKind == RefKind.None)
```
>sourceSignature.ReturnType.IsFunctionPointer() [](start = 63, length = 46)

Does this match the specification?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2364,6 +2448,61 @@ private void UpperBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+                var sourceParam = sourceSignature.Parameters[i];
+                var targetParam = targetSignature.Parameters[i];
+
+                if ((sourceParam.Type.IsReferenceType || sourceParam.Type.IsFunctionPointer()) && sourceParam.RefKind == RefKind.None)
```
>sourceParam.Type.IsFunctionPointer() [](start = 57, length = 36)

Does this match the specification?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2364,6 +2448,61 @@ private void UpperBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+                }
+            }
+
+            if ((sourceSignature.ReturnType.IsReferenceType || sourceSignature.ReturnType.IsFunctionPointer()) && sourceSignature.RefKind == RefKind.None)
```
>sourceSignature.ReturnType.IsFunctionPointer() [](start = 63, length = 46)

Does this match the specification?

333fred commented:
@AlekseyTs I believe I've addressed your feedback. I didn't see a review complete comment, so for reference your comments were on commit 3.

AlekseyTs commented:
@333fred It looks like there are some legitimate test failures.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -2069,6 +2219,115 @@ void M(bool b)
+        }
+
+        [Fact, WorkItem(50096, "https://github.com/dotnet/roslyn/issues/50096")]
+        public void FunctionPointerInference_ThroughMethodGroup()
```
>FunctionPointerInference_ThroughMethodGroup [](start = 20, length = 43)

I am assuming this test is added to cover changes made in commit 4. It feels like we need to add similar coverage for an address of and lambdas.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2067,6 +2101,61 @@ private void LowerBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+#nullable enable
+        private bool LowerBoundFunctionPointerTypeInference(TypeSymbol source, TypeSymbol target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            if (source is not FunctionPointerTypeSymbol { Signature: { } sourceSignature } || target is not FunctionPointerTypeSymbol { Signature: { } targetSignature })
```
>target is not FunctionPointerTypeSymbol { Signature: { } targetSignature } [](start = 94, length = 74)

I am not sure where we deal with the conversion part from the specification: "such that `U` has an implicit function pointer conversion to `delegate*<U2..Uk, U1>`"

333fred replied:
The specification no longer has that section.

---
In reply to: [558412661](https://github.com/dotnet/roslyn/pull/50249#discussion_r558412661) [](ancestors = 558412661)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2364,6 +2458,61 @@ private void UpperBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
+#nullable enable
+        private bool UpperBoundFunctionPointerTypeInference(TypeSymbol source, TypeSymbol target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            if (source is not FunctionPointerTypeSymbol { Signature: { } sourceSignature } || target is not FunctionPointerTypeSymbol { Signature: { } targetSignature })
```
>if (source is not FunctionPointerTypeSymbol { Signature: { } sourceSignature } || target is not FunctionPointerTypeSymbol { Signature: { } targetSignature }) [](start = 12, length = 157)

I am not sure where we deal with the conversion part from the specification: " `V` is a function pointer type which has an implicit function pointer conversion to `delegate*<U2..Uk, U1>`."

333fred replied:
The specification no longer has that section.

---
In reply to: [558415248](https://github.com/dotnet/roslyn/pull/50249#discussion_r558415248) [](ancestors = 558415248)

AlekseyTs commented:
Done with review pass (commit 4), tests are not looked at.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [761040006](https://github.com/dotnet/roslyn/pull/50249#issuecomment-761040006) [](ancestors = 761040006)

AlekseyTs commented:
>                      target.Type is FunctionPointerTypeSymbol { Signature: { ParameterCount: int targetParameterCount } targetSignature } &&

Should we check calling conventions and ref-ness too?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:1650 in 35ce5d0. [](commit_id = 35ce5d08d5f99cd998a617708c1c643b9eb953ab, deletion_comment = False)

AlekseyTs commented:
> addressed feedback.

@333fred It looks like there are some legitimate test failures.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -2069,6 +2219,183 @@ void M(bool b)
+                //     Test2<string, int>(0, &converter);
+                Diagnostic(ErrorCode.ERR_BadArgType, "&converter").WithArguments("2", "&method group", "System.Func<string, int>").WithLocation(8, 27)
+            );
+        }
```
>} [](start = 8, length = 1)

It looks like the change of "If Ui is not known to be a reference type then an exact inference is made" to "If Ui is not known to be a reference type or a function pointer type then an exact inference is made" in the specification enables the inference through arrays and generic types. Is this intentional? Are we testing that?

333fred replied:
The spec has been updated to address this.

---
In reply to: [558722993](https://github.com/dotnet/roslyn/pull/50249#discussion_r558722993) [](ancestors = 558722993)

AlekseyTs commented:
Done with review pass (commit 6)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [761278444](https://github.com/dotnet/roslyn/pull/50249#issuecomment-761278444) [](ancestors = 761278444)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -1621,6 +1650,13 @@ private bool ExactPointerInference(TypeWithAnnotations source, TypeWithAnnotatio
                      sourceParameterCount == targetParameterCount)
             {
+                if (sourceSignature.RefKind != targetSignature.RefKind
+                    || sourceSignature.ParameterRefKinds != targetSignature.ParameterRefKinds
```
>!= [](start = 57, length = 2)

What does this actually do? It looks like we are using SequenceEqual for lower/upper bound inference.

333fred replied:
Fixed and added a test to cover.

---
In reply to: [562997374](https://github.com/dotnet/roslyn/pull/50249#discussion_r562997374) [](ancestors = 562997374)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -1621,6 +1650,13 @@ private bool ExactPointerInference(TypeWithAnnotations source, TypeWithAnnotatio
             {
+                if (sourceSignature.RefKind != targetSignature.RefKind
+                    || sourceSignature.ParameterRefKinds != targetSignature.ParameterRefKinds
+                    || !callingConventionsEqual(sourceSignature, targetSignature))
```
>!callingConventionsEqual(sourceSignature, targetSignature) [](start = 23, length = 58)

Should we do the same check for upper/lower bound inference?

333fred replied:
Upper and lower bound inference do actual overload resolution, which handles refkind and calling convention mismatches.

---
In reply to: [562997572](https://github.com/dotnet/roslyn/pull/50249#discussion_r562997572) [](ancestors = 562997572)

AlekseyTs replied:
> Upper and lower bound inference do actual overload resolution, which handles refkind and calling convention mismatches.

I do not see LowerBoundFunctionPointerTypeInference and UpperBoundFunctionPointerTypeInference do overload resolution. What am I missing.

---
In reply to: [563941011](https://github.com/dotnet/roslyn/pull/50249#discussion_r563941011) [](ancestors = 563941011,562997572)

333fred replied:
> I do not see LowerBoundFunctionPointerTypeInference and UpperBoundFunctionPointerTypeInference do overload resolution. What am I missing?

It's done in phase 2 of inference, whereas this is all part of phase 1. I could add the checks here as well, but there won't be an observable output change.

---
In reply to: [564004088](https://github.com/dotnet/roslyn/pull/50249#discussion_r564004088) [](ancestors = 564004088,563941011,562997572)

AlekseyTs replied:
> It's done in phase 2 of inference, whereas this is all part of phase 1. 

This doesn't quite explain anything for me.

> I could add the checks here as well, but there won't be an observable output change.

This sounds suspicious to me. Could you please elaborate, pointing where the check would be performed instead?

---
In reply to: [564009761](https://github.com/dotnet/roslyn/pull/50249#discussion_r564009761) [](ancestors = 564009761,564004088,563941011,562997572)

333fred replied:
Hmm, actually, I believe in my testing I wasn't hitting the codepath because I was using address of method groups, which need to perform overload resolution (and will fail to do so in the presence of mismatches). I can hit it with variables of the appropriate pointer type, I'll add more testing.

---
In reply to: [564016180](https://github.com/dotnet/roslyn/pull/50249#discussion_r564016180) [](ancestors = 564016180,564009761,564004088,563941011,562997572)

333fred replied:
Tests added.

---
In reply to: [564040602](https://github.com/dotnet/roslyn/pull/50249#discussion_r564040602) [](ancestors = 564040602,564016180,564009761,564004088,563941011,562997572)

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [765812574](https://github.com/dotnet/roslyn/pull/50249#issuecomment-765812574) [](ancestors = 765812574)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -1631,6 +1665,32 @@ private bool ExactPointerInference(TypeWithAnnotations source, TypeWithAnnotatio
+                   && (sourceSignature.ParameterRefKinds.IsDefault, targetSignature.ParameterRefKinds.IsDefault) switch
+                   {
+                       (true, true) => true,
+                       (true, _) or (_, true) => false,
```
>_ [](start = 30, length = 1)

Explicitly stating ```false``` here is likely to improve readability.

AlekseyTs replied:
>_ [](start = 37, length = 1)

here as well

AlekseyTs commented:
Done with review pass (commit 9)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [767076222](https://github.com/dotnet/roslyn/pull/50249#issuecomment-767076222) [](ancestors = 767076222)

AlekseyTs reviewed (approved):
LGTM (commit 12), assuming CI is passing.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -1281,60 +1291,79 @@ private bool MethodGroupReturnTypeInference(Binder binder, BoundExpression sourc
             // SPEC:   yields a single method with return type U then a lower-bound
             // SPEC:   inference is made from U to Tb.
```
nit: Consider updating the spec comments to reflect changes.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -2067,6 +2132,54 @@ private void LowerBoundTypeArgumentInference(NamedTypeSymbol source, NamedTypeSy
         }
 
+#nullable enable
+        private bool LowerBoundFunctionPointerTypeInference(TypeSymbol source, TypeSymbol target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```
>LowerBoundFunctionPointerTypeInference [](start = 21, length = 38)

Adding new spec comments would help

jcouv reviewed (approved):
LGTM Thanks (iteration 12)

333fred commented:
@jcouv I'm going to follow up on your comment suggestions in master to get this in now.
